pkg/network/java/hotspot.go

```diff
@@ -0,0 +1,286 @@
+		cmd := append([]byte(c), 0)
+		if _, err := h.conn.Write(cmd); err != nil {
+			return err
+		}
+	}
+	if tailingNull {
+		if _, err := h.conn.Write([]byte{0}); err != nil {
+			return err
+		}
+	}
+
+	buf := make([]byte, 8192)
+	if _, err := h.conn.Read(buf); err != nil {
+		return err
+	}
+	returnCommand, returnCode, responseText := h.parseResponse(buf)
+
+	if returnCommand != 0 {
+		return fmt.Errorf("command '%s' return command %d code %d\n%s\n", cmd, returnCommand, returnCode, responseText)
+	}
+	return nil
+}
+
+// the (short) protocol is following
+//  o create a file .attach_pid%d
+//  o send a SIGQUIT signal
+//  o wait for socket file created by the java process
+func (h *Hotspot) attachJVMProtocol(uid int, gid int) error {
+	attachPIDPath := func(root string) string {
+		return fmt.Sprintf("%s/.attach_pid%d", root, h.nsPid)
+	}
+
+	attachPath := attachPIDPath(h.root + h.cwd)
+	hook, err := os.OpenFile(attachPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0660)
+	if err != nil {
+		return err
+	}
+	hook.Close()
+	// we don't check Chown() return error here as it can failed on some filesystem
+	_ = syscall.Chown(attachPath, uid, gid)
+	hookUID, _, ownerErr := getPathOwner(attachPath)
+	if err != nil || ownerErr != nil || hookUID != uint32(uid) {
+		// we failed to create the .attach_pid file in the process directory
+		// let's try in /tmp
+
+		// Note: some filesystem over write the owner when creating a file
+		// JVM doesn't like this
+		if ownerErr == nil {
+			os.Remove(attachPath)
+		}
+
+		attachPath = attachPIDPath(h.tmpPath())
+		hook, err = os.OpenFile(attachPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0660)
+		if err != nil {
+			return err
+		}
+		hook.Close()
+	}
+	defer os.Remove(attachPath)
+	if err != nil {
+		return err
+	}
+
+	process, _ := os.FindProcess(h.pid)
+	if err := process.Signal(syscall.SIGQUIT); err != nil {
+		return fmt.Errorf("process %d/%d SIGQUIT failed : %s", h.pid, h.nsPid, err)
+	}
+
+	h.socketPath = fmt.Sprintf("%s/.java_pid%d", h.tmpPath(), h.nsPid)
+	end := time.Now().Add(6 * time.Second)
+	for end.After(time.Now()) {
+		time.Sleep(200 * time.Millisecond)
+		if h.socketExists() {
+			return nil
+		}
+	}
+	return fmt.Errorf("the java process %d/%d didn't create a socket file", h.pid, h.nsPid)
+}
+
+// Attach() an agent to the hostspot, uid/gid must be accessible read-only by the targeted hotspot
+func (h *Hotspot) Attach(agent string, args string, uid int, gid int) error {
+
+	// ask JVM to create a socket to communicate
+	if err := h.attachJVMProtocol(uid, gid); err != nil {
+		return err
+	}
+
+	// copy the agent in the cwd of the process and change his owner/group
+	agentPath, agentCleanup, err := h.copyAgent(agent, uid, gid)
+	if err != nil {
+		return err
+	}
+	defer agentCleanup()
+
+	// connect and ask to load the agent .jar or .so
+	cleanConn, err := h.connect()
+	if err != nil {
+		return err
+	}
+	defer cleanConn()
+
+	var loadCommand string
+	isJar := strings.HasSuffix(filepath.Base(agent), ".jar")
```

Comment:
redundant call to filepathBase
```suggestion
	isJar := strings.HasSuffix(agent, ".jar")
```